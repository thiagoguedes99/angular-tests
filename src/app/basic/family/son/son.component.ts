import { Component, EventEmitter, Input, Output } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Client } from '../interfaces/client';

@Component({
  selector: 'app-son',
  standalone: true,
  imports: [CommonModule],
  templateUrl: './son.component.html',
  styleUrl: './son.component.scss'
})
export class SonComponent {
  @Input() client?: Client;
  @Output() onDeleteClient  = new EventEmitter();
  @Output() onClientUpdated = new EventEmitter<Client>();

  onDelete() {
    this.client = undefined;
    this.onDeleteClient.emit();
  }

  onChange( id: number ) {
    if ( !this.client ) return;

    // this.client.id = newId; // no usar
    this.client = { ...this.client, id };

    this.onClientUpdated.emit({ ...this.client })

  }
}
