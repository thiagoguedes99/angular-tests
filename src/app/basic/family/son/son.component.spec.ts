import { ComponentFixture, TestBed } from '@angular/core/testing';

import { SonComponent } from './son.component';

describe('SonComponent', () => {
  let component: SonComponent;
  let fixture: ComponentFixture<SonComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [SonComponent]
    })
    .compileComponents();

    fixture = TestBed.createComponent(SonComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should to match snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  it('should not view buttons if not has clinet', () => {
    const buttons = compiled.querySelectorAll('button');

    expect(buttons.length).toBe(0);
  });

  it('should to view buttons if has clinet', () => {
    component.client = { id: 1, name: 'thiago' };
    fixture.detectChanges();

    const buttons = compiled.querySelectorAll('button');

    expect(buttons.length).toBe(2);
  });

  it('should to match snapshot after has client', () => {
    component.client = { id: 1, name: 'thiago' };
    fixture.detectChanges();

    expect(compiled).toMatchSnapshot();
  });

  it('should to click button delete and to emit onDeleteClient', () => {
    component.client = { id: 1, name: 'thiago' };
    fixture.detectChanges();

    jest.spyOn(component.onDeleteClient, 'emit');

    const buttonDelete = compiled.querySelector('.btn-outline-danger');

    buttonDelete?.dispatchEvent( new Event('click') );

    expect(component.onDeleteClient.emit).toHaveBeenCalledTimes(1);
  });

  it('should to click button change id and to emit onClientUpdated', () => {
    component.client = { id: 1, name: 'thiago' };
    fixture.detectChanges();

    jest.spyOn(component.onClientUpdated, 'emit');

    const buttonUpdate = compiled.querySelector('.btn-outline-primary');

    buttonUpdate?.dispatchEvent( new Event('click') );

    expect(component.onClientUpdated.emit).toHaveBeenCalledTimes(1);
    expect(component.onClientUpdated.emit).toHaveBeenCalledWith({ id: 5, name: 'thiago' });
  });

  it('should not to emit onClientUpdated event if not has client', () => {
    component.client = undefined;
    jest.spyOn(component.onClientUpdated, 'emit');

    component.onChange(10);

    expect(component.onClientUpdated.emit).toHaveBeenCalledTimes(0);
  });

  it('should to emit onClientUpdated event with newv values client', () => {
    component.client = undefined;
    jest.spyOn(component.onClientUpdated, 'emit');

    component.onChange(10);

    expect(component.onClientUpdated.emit).toHaveBeenCalledTimes(0);

    component.client = { id: 1, name: 'thiago' };
    component.onChange(10);

    expect(component.onClientUpdated.emit).toHaveBeenCalledTimes(1);
    expect(component.onClientUpdated.emit).toHaveBeenCalledWith({ id: 10, name: 'thiago' });
  });

  it('should to emit onDeleteClient event and resert values client in onDelete function', () => {
    component.client = { id: 1, name: 'thiago' };

    expect(component.client).toEqual({ id: 1, name: 'thiago' });

    jest.spyOn(component.onDeleteClient, 'emit');

    component.onDelete();

    expect(component.client).toEqual(undefined);
    expect(component.onDeleteClient.emit).toHaveBeenCalledTimes(1);
  });
});
