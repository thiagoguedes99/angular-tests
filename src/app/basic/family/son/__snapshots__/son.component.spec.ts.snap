// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SonComponent should to match snapshot 1`] = `
<div
  id="root1"
>
  <h5>
    Hijo
  </h5>
  <code>
    
  </code>
  <br />
  <br />
  
  
</div>
`;

exports[`SonComponent should to match snapshot after has client 1`] = `
<div
  id="root4"
>
  <h5>
    Hijo
  </h5>
  <code>
    {
  "id": 1,
  "name": "thiago"
}
  </code>
  <br />
  <br />
  <button
    class="btn btn-outline-primary"
    data-test="btn-id"
  >
     Cambiar ID

  </button>
  
  <button
    class="btn btn-outline-danger"
    data-test="btn-delete"
  >
     Borrar cliente

  </button>
  
</div>
`;
